docs/tutorial.rst

```diff
@@ -306,17 +318,303 @@ trees. In this example we incorrectly inferred that 0 coalesces with the
+library.
+
++++++++++++++
+Reading a VCF
++++++++++++++
+
+A common way to store genetic variation data is as a VCF (Variant Call Format) file.
+The `CYVCF2 library <https://github.com/brentp/cyvcf2>`_ provides a relatively fast
+Python interface for reading VCF files. The following script demonstrates how to infer a
+tree sequence from a simple VCF file. In particular the function below named
+``add_diploid_sites`` illustrates how to iterate over the variants in a ``CYVCF2.VCF``
+object and add them to a ``tsinfer`` sample data file, while printing out the progression
+through the file in number of base pairs. This particular script assumes that the VCF
+describes a set of phased, diploid individuals, and has the ancestral state present in
+the INFO field of the VCF (otherwise it takes the reference allele as the ancestral
+state). For example data, we use a publicly available VCF file of the genetic variants
+from chromosome 24 of ten Norwegian and French house sparrows, *Passer domesticus*
+(thanks to Mark Ravinet for the data file):
 
-    1. Add documentation links for msprime above so we can explain tree
-       sequences there.
+.. code-block:: python
 
-++++++++++++
-Data example
-++++++++++++
+    import cyvcf2
+    import tsinfer
+    import tqdm
+
+    def add_diploid_sites(vcf, samples):
+        """
+        Loop over sites in the vcf and add them to the samples object, reordering the
+        alleles to put the ancestral allele first, if it is available.
+        """
+        progressbar = tqdm.tqdm(total=samples.sequence_length, desc="Read VCF", unit='bp')
+        pos = 0
+        for variant in vcf:
+            if pos and pos == variant.POS:
+                raise ValueError("Duplicate positions for variant at position", pos)
+            else:
+                progressbar.update(variant.POS - pos)
+                pos = variant.POS
+            if any([not phased for _, _, phased in variant.genotypes]):
+                raise ValueError("Unphased genotypes for variant at position", pos)
+            alleles = [variant.REF] + variant.ALT
+            # Sort with the ancestral allele first if we have one (otherwise use REF)
+            ancestral = variant.INFO.get('AA', variant.REF)
+            all_alleles = [variant.REF] + variant.ALT
+            # Ancestral state must be first in the allele list.
+            alleles = [ancestral] + list(set(all_alleles) - {ancestral})
+            allele_index = {old_index: alleles.index(allele)
+                for old_index, allele in enumerate(all_alleles)}
+            # Map original allele indexes to their indexes in the new alleles list.
+            genotypes = [allele_index[old_index]
+                for row in variant.genotypes for old_index in row[0:2]]
+            samples.add_site(pos, genotypes=genotypes, alleles=alleles)
+        progressbar.close()
+
+    def chromosome_length(vcf):
+        try:
```

Comment:
This is too general --- we're only writing a script to support just this one VCF file. Every bell and whistle that's added makes it that bit more difficult for people to understand, and we want to be totally and absolutely explicit that this is just a **guide** for writing your own parser.

Reply:
See below re comments about setting the sequence_length, which I do think we want to do (and encourage), even in this basic guide version.

Reply:
I disagree - I think we should be explicit about what we expect the VCF to contain and to write code to process only that. Having extra stuff in there just gives a false sense of generality.

Reply:
But that's what I mean: this particular VCF contains the sequence length (which is different from the last variant position), and I think we want to pass this as the `sequence_length` parameter when we make the SampleData file. It's merely a question of whether we have a function that does this, or if we pass it in direct as vcf.seqlen[0] and don't bother about ensuring that our script "fails noisily" on a bad (multi-length) VCF.